### ISSUE TITLE
Overriding packages in the builtin repository changed semantic from v0.16 to v0.17

### ISSUE THREAD

**Issue by npf** [LLNL/spack#28747]
### Steps to reproduce

I'm trying to use the overriding of packages as described in https://spack.readthedocs.io/en/latest/repositories.html#overriding-built-in-packages, with the following setup:
- the builtin repo contains a package with the most recent versions (e.g. zlib 1.2.11)
- a local repo contains the same package but only an old version (e.g. zlib 1.2.8)

localrepo is added to the repo list, thus it has priority over builtin. Since I want the latest version of the package, I run `spack install builtin.zlib@1.2.11`, but it does not behave as expected with v0.17.0 and v0.17.1, while it's ok with v0.16.3

### Error message

```console
$ spack install builtin.zlib@1.2.11
==> Warning: Missing a source id for zlib@1.2.11
==> Installing zlib-1.2.11-ysprwfsnfe2ybpqg4r3tkbkvyavgmjne
==> No binary for zlib-1.2.11-ysprwfsnfe2ybpqg4r3tkbkvyavgmjne found: installing from source
==> Warning: There is no checksum on file to fetch zlib@1.2.11 safely.
==>   Fetch anyway? [y/N] y
==> Fetching https://zlib.net/fossils/zlib-1.2.11.tar.gz
==> No patches needed for zlib
==> zlib: Executing phase: 'install'
==> zlib: Successfully installed zlib-1.2.11-ysprwfsnfe2ybpqg4r3tkbkvyavgmjne
  Fetch: 2.40s.  Build: 3.33s.  Total: 5.73s.
[+] /path/to/spack/opt/spack/linux-debian10-skylake/gcc-8.3.0/zlib-1.2.11-ysprwfsnfe2ybpqg4r3tkbkvyavgmjne
$ spack find -N
==> 2 installed packages
-- linux-debian10-skylake / gcc@8.3.0 ---------------------------
spack-localrepo.zlib@1.2.8  spack-localrepo.zlib@1.2.11
```
→ zlib@1.2.11 is not fetched from builtin

While it was ok with v0.16.3.
```console
$ spack install builtin.zlib@1.2.11
==> Installing zlib-1.2.11-35k2t3imxwoldp42uymipncstsisqeqt
==> No binary for zlib-1.2.11-35k2t3imxwoldp42uymipncstsisqeqt found: installing from source
==> Using cached archive: /home/neyron/scm/spack-v0.16.2/var/spack/cache/_source-cache/archive/c3/c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1.tar.gz
==> zlib: Executing phase: 'install'
[+] /path/to/spack/opt/spack/linux-debian10-skylake/gcc-8.3.0/zlib-1.2.11-35k2t3imxwoldp42uymipncstsisqeqt
$ spack find -N
==> 2 installed packages
-- linux-debian10-skylake / gcc@8.3.0 ---------------------------
builtin.zlib@1.2.11  spack-localrepo.zlib@1.2.8
```
→ zlib@1.2.11 is fetched from builtin

### Information on your system

* **Spack:** 0.17.1
* **Python:** 3.7.3
* **Platform:** linux-debian10-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I would also see how somebody could claim the behavior is correct in v0.17 and uncorrect in v0.16 - if we consider the local repository as an "overlay" :thinking: One way or the other we probably need to add docs on the expected behavior.

**Comment by alalazo** [comment]
Duplicate of #35536

**Comment by alalazo** [comment]
Closing this one as a duplicate let's follow in the other issue. For what is worth I raised the impact of #35536 due to how many times this has been reported :slightly_frowning_face: Hopefully we'll have a fix for v0.21.0 in November
